File: sdk/dart/lib/src/realtime_service.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'client.dart';
import 'types.dart';

/// One realtime change notification for a record.
class RealtimeEvent {
  final String action; // "create" | "update" | "delete"
  final RecordModel record;

  const RealtimeEvent({required this.action, required this.record});

  factory RealtimeEvent.fromJson(Map<String, dynamic> json) => RealtimeEvent(
        action: json['action'] as String? ?? '',
        record: RecordModel.fromJson(json['record'] as Map<String, dynamic>? ?? const {}),
      );
}

typedef RealtimeCallback = void Function(RealtimeEvent event);

/// Unsubscribes a single [RealtimeService.subscribe] call.
typedef Unsubscribe = void Function();

/// Live record change subscriptions over Server-Sent Events. Subscribe to a
/// bare collection name for every change, or `"collection/recordId"` for
/// just one record.
///
/// ```dart
/// final unsubscribe = await cb.realtime.subscribe('posts', (e) {
///   print('${e.action} ${e.record.id}');
/// });
/// // later
/// unsubscribe();
/// ```
///
/// Dart has no built-in `EventSource` outside a browser context, so this
/// opens the stream itself via `http.Client().send()` and hand-parses the
/// `event:`/`data:` line protocol.
class RealtimeService {
  final CratebaseClient client;

  http.Client? _sseClient;
  StreamSubscription<String>? _subscription;
  String _clientId = '';
  Completer<void>? _connectCompleter;
  final Map<String, Set<RealtimeCallback>> _subscriptions = {};

  RealtimeService(this.client);

  Future<Unsubscribe> subscribe(String topic, RealtimeCallback callback) async {
    _subscriptions.putIfAbsent(topic, () => {}).add(callback);
    await _ensureConnected();
    await _syncSubscriptions();
    return () => unsubscribe(topic, callback);
  }

  Future<void> unsubscribe(String topic, [RealtimeCallback? callback]) async {
    final set = _subscriptions[topic];
    if (set == null) return;
    if (callback != null) {
      set.remove(callback);
    } else {
      set.clear();
    }
    if (set.isEmpty) _subscriptions.remove(topic);

    if (_subscriptions.isEmpty) {
      disconnect();
    } else {
      await _syncSubscriptions();
    }
  }

  void disconnect() {
    _subscription?.cancel();
    _subscription = null;
    _sseClient?.close();
    _sseClient = null;
    _clientId = '';
    _connectCompleter = null;
  }

  Future<void> _ensureConnected() {
    final existing = _connectCompleter;
    if (existing != null) return existing.future;

    final completer = Completer<void>();
    _connectCompleter = completer;

    final httpClient = http.Client();
    _sseClient = httpClient;

    final uri = Uri.parse('${client.baseUrl}/api/realtime');
    final request = http.Request('GET', uri);
    request.headers['accept'] = 'text/event-stream';

    unawaited(_runStream(httpClient, request, completer));

    return completer.future;
  }

  Future<void> _runStream(
    http.Client httpClient,
    http.Request request,
    Completer<void> completer,
  ) async {
    try {
      final response = await httpClient.send(request);
      if (response.statusCode != 200) {
        throw StateError('failed to connect to /api/realtime: HTTP ${response.statusCode}');
      }

      var eventName = 'message';
      final dataBuffer = StringBuffer();

      void dispatch() {
        final dataStr = dataBuffer.toString();
        dataBuffer.clear();
        final name = eventName;
        eventName = 'message';
        if (dataStr.isEmpty) return;

        dynamic parsed;
        try {
          parsed = jsonDecode(dataStr);
        } catch (_) {
          return;
        }
        if (parsed is! Map) return;

        if (name == 'PB_CONNECT') {
          final clientId = parsed['clientId'];
          if (clientId is String) {
            _clientId = clientId;
            if (!completer.isCompleted) completer.complete();
          }
          return;
        }

        if (parsed['action'] is! String || parsed['record'] is! Map) return;
        final event = RealtimeEvent.fromJson(parsed.cast<String, dynamic>());
        final topics = [
          event.record.collectionName,
          '${event.record.collectionName}/${event.record.id}',
        ];
        for (final topic in topics) {
          final callbacks = _subscriptions[topic];
          if (callbacks == null) continue;
          for (final cb in Set<RealtimeCallback>.from(callbacks)) {
            cb(event);
          }
        }
      }

      _subscription = response.stream
          .transform(utf8.decoder)
          .transform(const LineSplitter())
          .listen(
        (line) {
          if (line.isEmpty) {
            dispatch();
            return;
          }
          if (line.startsWith(':')) return; // SSE comment/keep-alive
          if (line.startsWith('event:')) {
            eventName = line.substring(6).trim();
          } else if (line.startsWith('data:')) {
            if (dataBuffer.length > 0) dataBuffer.write('\n');
            dataBuffer.write(line.substring(5).trim());
          }
          // `id:`/`retry:` lines are ignored — the server doesn't rely on them.
        },
        onError: (Object error) {
          if (!completer.isCompleted) completer.completeError(error);
          disconnect();
        },
        onDone: () {
          if (!completer.isCompleted) {
            completer.completeError(StateError('connection to /api/realtime closed before PB_CONNECT'));
          }
          disconnect();
        },
        cancelOnError: true,
      );
    } catch (error) {
      if (!completer.isCompleted) completer.completeError(error);
      disconnect();
    }
  }

  Future<void> _syncSubscriptions() async {
    if (_clientId.isEmpty) return;
    await client.send(
      '/api/realtime',
      method: 'POST',
      body: {'clientId': _clientId, 'subscriptions': _subscriptions.keys.toList()},
    );
  }
}

File: sdk/go/realtime_service.go
package cratebase

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"sync"
)

// Event is a realtime record change notification.
type Event struct {
	Action string      `json:"action"` // "create" | "update" | "delete"
	Record RecordModel `json:"record"`
}

// Callback receives realtime events for a subscribed topic.
type Callback func(Event)

// RealtimeService provides live record change subscriptions over
// Server-Sent Events. Subscribe to a bare collection name for every
// change, or "collection/recordId" for just one record.
//
//	unsubscribe, err := client.Realtime.Subscribe("posts", func(e cratebase.Event) {
//		fmt.Println(e.Action, e.Record.ID())
//	})
//	// later
//	unsubscribe()
type RealtimeService struct {
	client *Client

	mu            sync.Mutex
	clientID      string
	connected     bool
	connCh        chan error // set while a connection attempt is in flight
	stop          chan struct{}
	subscriptions map[string][]subscriberEntry
	nextID        uint64
}

type subscriberEntry struct {
	id       uint64
	callback Callback
}

func newRealtimeService(client *Client) *RealtimeService {
	return &RealtimeService{
		client:        client,
		subscriptions: make(map[string][]subscriberEntry),
	}
}

// Subscribe registers callback for topic, opening the SSE connection if
// it isn't already open, and returns an unsubscribe func.
func (r *RealtimeService) Subscribe(topic string, callback Callback) (func(), error) {
	r.mu.Lock()
	r.nextID++
	id := r.nextID
	r.subscriptions[topic] = append(r.subscriptions[topic], subscriberEntry{id: id, callback: callback})
	r.mu.Unlock()

	if err := r.ensureConnected(); err != nil {
		return nil, err
	}
	if err := r.syncSubscriptions(); err != nil {
		return nil, err
	}

	return func() {
		r.removeSubscriber(topic, id)
	}, nil
}

func (r *RealtimeService) removeSubscriber(topic string, id uint64) {
	r.mu.Lock()
	entries := r.subscriptions[topic]
	for i, e := range entries {
		if e.id == id {
			entries = append(entries[:i], entries[i+1:]...)
			break
		}
	}
	if len(entries) == 0 {
		delete(r.subscriptions, topic)
	} else {
		r.subscriptions[topic] = entries
	}
	empty := len(r.subscriptions) == 0
	r.mu.Unlock()

	if empty {
		r.Disconnect()
	} else {
		_ = r.syncSubscriptions()
	}
}

// Unsubscribe removes every callback registered for topic and stops the
// stream if nothing remains subscribed.
func (r *RealtimeService) Unsubscribe(topic string) error {
	r.mu.Lock()
	delete(r.subscriptions, topic)
	empty := len(r.subscriptions) == 0
	r.mu.Unlock()

	if empty {
		r.Disconnect()
		return nil
	}
	return r.syncSubscriptions()
}

// Disconnect closes the SSE stream, if open.
func (r *RealtimeService) Disconnect() {
	r.mu.Lock()
	if r.stop != nil {
		close(r.stop)
		r.stop = nil
	}
	r.connected = false
	r.clientID = ""
	r.mu.Unlock()
}

func (r *RealtimeService) ensureConnected() error {
	r.mu.Lock()
	if r.connected {
		r.mu.Unlock()
		return nil
	}
	if r.connCh != nil {
		ch := r.connCh
		r.mu.Unlock()
		return <-ch
	}
	connCh := make(chan error, 1)
	stop := make(chan struct{})
	r.connCh = connCh
	r.stop = stop
	r.mu.Unlock()

	go r.readLoop(connCh, stop)

	return <-connCh
}

func (r *RealtimeService) readLoop(connCh chan error, stop chan struct{}) {
	req, err := http.NewRequest(http.MethodGet, r.client.BaseURL+"/api/realtime", nil)
	if err != nil {
		connCh <- err
		return
	}
	req.Header.Set("Accept", "text/event-stream")

	httpClient := r.client.HTTPClient
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		connCh <- fmt.Errorf("cratebase: connecting to /api/realtime: %w", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		connCh <- fmt.Errorf("cratebase: /api/realtime returned status %d", resp.StatusCode)
		return
	}

	go func() {
		<-stop
		resp.Body.Close()
	}()

	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	var eventName string
	var dataLines []string
	connected := false

	flush := func() {
		if len(dataLines) == 0 {
			eventName = ""
			return
		}
		data := strings.Join(dataLines, "\n")
		dataLines = nil
		name := eventName
		eventName = ""

		switch name {
		case "PB_CONNECT":
			var payload struct {
				ClientID string `json:"clientId"`
			}
			if err := json.Unmarshal([]byte(data), &payload); err == nil && payload.ClientID != "" {
				r.mu.Lock()
				r.clientID = payload.ClientID
				r.connected = true
				r.connCh = nil
				r.mu.Unlock()
				if !connected {
					connected = true
					connCh <- nil
				}
			}
		default: // "message" or unnamed
			var event Event
			if err := json.Unmarshal([]byte(data), &event); err != nil || event.Record == nil {
				return
			}
			r.dispatch(event)
		}
	}

	for scanner.Scan() {
		select {
		case <-stop:
			return
		default:
		}

		line := scanner.Text()
		switch {
		case line == "":
			flush()
		case strings.HasPrefix(line, "event:"):
			eventName = strings.TrimSpace(strings.TrimPrefix(line, "event:"))
		case strings.HasPrefix(line, "data:"):
			dataLines = append(dataLines, strings.TrimSpace(strings.TrimPrefix(line, "data:")))
		case strings.HasPrefix(line, ":"):
			// comment/keepalive, ignore
		}
	}

	if !connected {
		connCh <- fmt.Errorf("cratebase: /api/realtime stream closed before connecting")
	}

	r.mu.Lock()
	r.connected = false
	r.clientID = ""
	r.mu.Unlock()
}

func (r *RealtimeService) dispatch(event Event) {
	collectionName := event.Record.CollectionName()
	recordID := event.Record.ID()
	topics := []string{collectionName, collectionName + "/" + recordID}

	r.mu.Lock()
	var callbacks []Callback
	for _, topic := range topics {
		for _, e := range r.subscriptions[topic] {
			callbacks = append(callbacks, e.callback)
		}
	}
	r.mu.Unlock()

	for _, cb := range callbacks {
		cb(event)
	}
}

func (r *RealtimeService) syncSubscriptions() error {
	r.mu.Lock()
	clientID := r.clientID
	topics := make([]string, 0, len(r.subscriptions))
	for topic := range r.subscriptions {
		topics = append(topics, topic)
	}
	r.mu.Unlock()

	if clientID == "" {
		return nil
	}
	return r.client.Send("/api/realtime", SendOptions{
		Method: "POST",
		Body:   map[string]any{"clientId": clientID, "subscriptions": topics},
	}, nil)
}
